rfcs/2020-06-12-2768-batch-and-buffer-rework.md

```diff
@@ -0,0 +1,133 @@
+   both. Worse, this is hard-coded into the sink, so that the choice of
+   which one is allowed is made by Vector.
+2. For sinks that limit batch sizes in bytes, a buffer is considered
+   ready to send when the size reaches *or exceeds* the given size
+   limit. If the size limit is close to the maximum allow request size
+   for a remote service, this will routinely lead to problems when the
+   flow rate is high enough to avoid triggering a batch timeout.
+
+## Guide-Level Proposal
+
+For sinks that use the `BatchSink` framework (`PartitionBatchSink` is
+similar), events move along the following path:
+
+1. Each event is optionally encoded into a more suitable format for
+   buffering (using `with_flat_map` on the batch sink, ref `type
+   Batch::Input`).
+2. The encoded item is pushed into a buffer (in
+   `BatchSink::start_send`).
+3. When that buffer is full, as determined by being at or above the
+   configured batch size, it is collected into a batch (in
+   `BatchSink::poll_complete`).
+4. The batch is turned into a request (in `trait Service::call`) which
+   will do the final encoding and serialization.
+
+There exist the following buffer types:
+
+- `Buffer` offers (gzip) compression, but requires the events be
+  serialized before insertion and that serialization must include a
+  terminator. Events are stored and output as one large `Vec<u8>`, which
+  allows for easy computation of both sizes.
+- `JsonArrayBuffer` internally serializes events into "raw" JSON, which
+  may be then added to another structure after output. Events are stored
+  as `serde_json::value::RawValue` which is a newtype over a boxed
+  string, which allows for maintaining a running total of both sizes.
+- `MetricBuffer` stores metrics in their structured form, without
+  encoding. Since the target serialization is unknown, this buffer
+  cannot know the total byte size.
+- `PartitionBuffer<T, K>` is a meta-buffer, that partitions events into
+  multiple buffers of type `T`.
+- `Vec<T>` can only count events, not bytes, as it is agnostic to the
+  size of `T`.
+
+We want to get to the point where as many batching sinks as possible can
+be configured with either a byte limit or an event limit, or
+both. Additionally, the batch limits must be inviolable, so that no
+requests are ever sent that exceed size limits. To accomplish this the
+following is necessary:
+
+### Unified Batch Configuration
+
+The split configuration types for bytes and events will be rolled into a
+single configuration combining both attributes. Batch sizes will be
+limited to which ever maximum is reached first.
+
+The `max_size` element is retained for backwards compatibility, and each
+sink will use it as a default for the maximum bytes or events depending
+on the previous configuration mode. If both `max_size` and either of
+`max_bytes` or `max_events` are set, Vector will produce a configuration
+error indicating a conflict in the configuration.
+
+```rust
+struct BatchConfig {
```

Comment:
It’s possible to use `enum` instead of a `struct` here to avoid ambiguity at the type system level. Serde is then capable to deserialize the enum to a specific variant based in which fields are present (as long as the names don’t repeat), or emitting a parsing error.
https://serde.rs/enum-representations.html

Reply:
I'm not sure it can. Any or all three of `max_bytes`, `max_events`, and `timeout_secs` may be set in this struct, or alternately any or all two of `max_size` and `timeout_secs` may be set. It may be possible to use an `enum` within this `struct`, but it's not clear to me how as it isn't a strict either-or choice.